truth_set/python/structure/solution-approach/semantic-duplication/services/orderService.py

```diff
@@ -0,0 +1,32 @@
+from datetime import datetime
+
+class OrderService:
+    def __init__(self):
+        self.orders = []
+    
+    def add_order(self, product):
+        new_order = {
+            'order_id': len(self.orders) + 1,
+            'product': product,
+            'order_date': datetime.now()
+        }
+        self.orders.append(new_order)
+        formatted_date = self.format_date(new_order['order_date'], 'DD-MM-YYYY')
+        print(f"Order Added: {new_order['product']}, Order Date: {formatted_date}")
+    
+    def list_orders(self):
+        for order in self.orders:
+            formatted_date = self.format_date(order['order_date'])
+            print(f"Order ID: {order['order_id']}, Product: {order['product']}, Order Date: {formatted_date}")
```

Comment:
<div>


<div id="suggestion">
<div id="issue"><b>Consider base service class for common functionality</b></div>
<div id="fix">

The `OrderService` class appears to have similar functionality to `UserService` for adding and listing items. Consider creating a base service class to avoid semantic duplication. Similar issues were also found in:
- truth_set/python/structure/solution-approach/semantic-duplication/services/orderService.py (line 22-32)
- truth_set/python/structure/solution-approach/semantic-duplication/services/userService.py (line 15-19)
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<blockquote>Check the AI-generated fix before applying</blockquote>
<div id="code">

```
 @@ -1,20 +1,30 @@
 +class BaseService:
 +    def __init__(self):
 +        self.items = []
 +
 +    def add_item(self, item_data):
 +        self.items.append(item_data)
 +
 +    def list_items(self):
 +        for item in self.items:
 +            self._print_item(item)
 +
 +    def _print_item(self, item):
 +        pass
 +
 class OrderService(BaseService):
     def __init__(self):
 -        self.orders = []
 +        super().__init__()

     def add_order(self, product):
         new_order = {
             'order_id': len(self.orders) + 1,
             'product': product,
             'order_date': datetime.now()
         }
 -        self.orders.append(new_order)
 +        self.add_item(new_order)

 -    def list_orders(self):
 -        for order in self.orders:
 -            formatted_date = self.format_date(order['order_date'])
 -            print(f"Order ID: {order['order_id']}, Product: {order['product']}, Order Date: {formatted_date}")
 +    def _print_item(self, order):
 +        formatted_date = self.format_date(order['order_date'])
 +        print(f"Order ID: {order['order_id']}, Product: {order['product']}, Order Date: {formatted_date}")
```




```
 @@ -1,20 +1,30 @@
 +class BaseService:
 +    def __init__(self):
 +        self.items = []
 +
 +    def add_item(self, item_data):
 +        self.items.append(item_data)
 +
 +    def list_items(self):
 +        for item in self.items:
 +            self._print_item(item)
 +
 +    def _print_item(self, item):
 +        pass
 +
 class OrderService(BaseService):
     def __init__(self):
 -        self.orders = []
 +        super().__init__()

     def add_order(self, product):
         new_order = {
             'order_id': len(self.orders) + 1,
             'product': product,
             'order_date': datetime.now()
         }
 -        self.orders.append(new_order)
 +        self.add_item(new_order)

 -    def list_orders(self):
 -        for order in self.orders:
 -            formatted_date = self.format_date(order['order_date'])
 -            print(f"Order ID: {order['order_id']}, Product: {order['product']}, Order Date: {formatted_date}")
 +    def _print_item(self, order):
 +        formatted_date = self.format_date(order['order_date'])
 +        print(f"Order ID: {order['order_id']}, Product: {order['product']}, Order Date: {formatted_date}")
```

</div>
</details>
</div>


<small><i>Code Review Run <a href=https://github.com/rkj6124/bito-test/pull/1#issuecomment-2520822763>#7b761a</a></i></small>
</div>

---
Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged